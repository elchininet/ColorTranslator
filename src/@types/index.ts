export type NumberOrString = number | string;

export interface HEXObject {
    r: string;
    g: string;
    b: string;
    a?: string;
}

export interface RGBObject {
    r: number;
    g: number;
    b: number;
    a?: number;
}

export interface RGBObjectGeneric {
    r: NumberOrString;
    g: NumberOrString;
    b: NumberOrString;
    a?: NumberOrString;
}

export interface HSLObject {
    h: number;
    s: number;
    l: number;
    a?: number;
}

export interface HSLObjectGeneric {
    h: number;
    s: NumberOrString;
    l: NumberOrString;
    a?: number;
}

export interface CMYKObject {
    c: number;
    m: number;
    y: number;
    k: number;
    a?: number;
}

export interface CMYKObjectGeneric {
    c: NumberOrString;
    m: NumberOrString;
    y: NumberOrString;
    k: NumberOrString;
    a?: number;
}

export interface RYBObject {
    r: number;
    y: number;
    b: number;
    a?: number;
}

export type RGYBObject = RGBObject | RYBObject;

export type Color = RGBObjectGeneric | HSLObjectGeneric | CMYKObjectGeneric;
export type ColorWithoutCMYK = RGBObjectGeneric | HSLObjectGeneric;

export type ColorInput = string | Color;
export type ColorInputWithoutCMYK = string | ColorWithoutCMYK;
export type HEXOutput = string | HEXObject;
export type RGBOutput = string | RGBObject;
export type HSLOutput = string | HSLObject;
export type CMYKOutput = string | CMYKObject;
export type ColorOutput = HEXOutput | RGBOutput | HSLOutput;

export interface ObjectProps<T> {
    [key: string]: T;
}

export interface Options {
    decimals: number;
    legacyCSS: boolean;
    spacesAfterCommas: boolean;
}

export type InputOptions = Partial<Options>;