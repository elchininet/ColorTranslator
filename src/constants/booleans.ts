export const DEFAULT_COLOR_LEVEL_4 = false;