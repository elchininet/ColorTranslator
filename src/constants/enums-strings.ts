import { Harmony, Mix } from './enums';

export type HarmonyString = `${Harmony}`;
export type MixString = `${Mix}`;