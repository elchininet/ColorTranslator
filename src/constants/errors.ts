export const ERRORS = {
    NOT_ACCEPTED_STRING_INPUT: 'The provided string color doesn\'t have a correct format',
    NOT_ACCEPTED_OBJECT_INPUT: 'The provided color object doesn\'t have the proper keys or format',
    NOT_A_VALID_RELATIVE_COLOR: 'The provided relative color is not valid'
};