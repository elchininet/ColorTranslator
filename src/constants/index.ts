export * from './numbers';
export * from './enums';
export * from './regexps';
export * from './errors';
