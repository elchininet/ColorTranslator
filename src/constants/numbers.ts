export const MAX_DECIMALS = 6;
export const MIN_DECIMALS = 2;
