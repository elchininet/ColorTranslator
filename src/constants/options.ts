import {
    Options,
    AnglesUnitEnum,
    ColorUnitEnum,
    CMYKFunctionEnum
} from '@types';
import { MAX_DECIMALS } from './numbers';

export const DEFAULT_OPTIONS: Options = {
    decimals: MAX_DECIMALS,
    legacyCSS: false,
    spacesAfterCommas: false,
    anglesUnit: AnglesUnitEnum.NONE,
    rgbUnit: ColorUnitEnum.NONE,
    labUnit: ColorUnitEnum.NONE,
    cmykUnit: ColorUnitEnum.PERCENT,
    alphaUnit: ColorUnitEnum.NONE,
    cmykFunction: CMYKFunctionEnum.DEVICE_CMYK
};