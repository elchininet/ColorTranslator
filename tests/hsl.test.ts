import { ColorTranslator } from '../src';
import { COLORS } from './tests.constants';

const HSL_REGEXP = /^(hsla?\()(\d+)([^)]*\))$/;

COLORS.forEach((item): void => {

    describe(`HSL tests for ${item.hsl}`, (): void => {

        const hsl = { ...item.hslObject };
        const hsla = { ...item.hslaObject };
        const hslPercentages = { ...hsl, s: hsl.s + '%', l: hsl.l + '%' };
        const hslaPercentages = { ...hsla, s: hsla.s + '%', l: hsla.l + '%' };
        const hslBigHue = { ...hsl, h: hsl.h + 360 };
        const hslaBigHue = { ...hsl, h: hsl.h + 360 };
        const hslLowHue = { ...hsl, h: hsl.h - 360 };
        const hslaLowHue = { ...hsl, h: hsl.h - 360 };
        const hslDeg = item.hsl.replace(HSL_REGEXP, '$1$2deg$3');
        const hslGrad = item.hsl.replace(HSL_REGEXP, '$1$2grad$3');
        const hslRadians = item.hsl.replace(HSL_REGEXP, (__match: string, start: string, number: string, end: string): string => {
            return `${start}${Math.round((+number * Math.PI / 180) * 10000) / 10000}rad${end}`;
        });
        const hslTurns = item.hsl.replace(HSL_REGEXP, (__match: string, start: string, number: string, end: string): string => {
            return `${start}${Math.round(+number / 360 * 10000) / 10000}turn${end}`;
        });
        const options = { decimals: 0 };
        const legacyOptions = { ...options, legacyCSS: true };

        it(`HSL with percentages ${JSON.stringify(hslPercentages)}`, (): void => {

            expect(ColorTranslator.toRGB(hslPercentages, options)).toBe(item.rgb);
            expect(ColorTranslator.toRGB(hslPercentages, legacyOptions)).toBe(item.rgbLegacy);
            expect(ColorTranslator.toRGBObject(hslPercentages, options)).toMatchObject(item.rgbObject);

            expect(ColorTranslator.toHSL(hslPercentages, options)).toBe(item.hsl);
            expect(ColorTranslator.toHSL(hslPercentages, legacyOptions)).toBe(item.hslLegacy);
            expect(ColorTranslator.toHSLObject(hslPercentages, options)).toMatchObject(item.hslObject);

            expect(ColorTranslator.toRGB(hslPercentages)).toMatchSnapshot();
            expect(ColorTranslator.toHSL(hslPercentages)).toMatchSnapshot();

        });

        it(`HSLA with percentages ${JSON.stringify(hslaPercentages)}`, (): void => {

            expect(ColorTranslator.toRGBA(hslaPercentages, options)).toBe(item.rgba);
            expect(ColorTranslator.toRGBA(hslaPercentages, legacyOptions)).toBe(item.rgbaLegacy);
            expect(ColorTranslator.toRGBAObject(hslaPercentages, options)).toMatchObject(item.rgbaObject);

            expect(ColorTranslator.toHSLA(hslaPercentages, options)).toBe(item.hsla);
            expect(ColorTranslator.toHSLA(hslaPercentages, legacyOptions)).toBe(item.hslaLegacy);
            expect(ColorTranslator.toHSLAObject(hslaPercentages, options)).toMatchObject(item.hslaObject);

            expect(ColorTranslator.toRGBA(hslPercentages)).toMatchSnapshot();
            expect(ColorTranslator.toHSLA(hslPercentages)).toMatchSnapshot();

        });

        it(`HSL big hue ${JSON.stringify(hslBigHue)}`, (): void => {

            expect(ColorTranslator.toRGB(hslBigHue, options)).toBe(item.rgb);
            expect(ColorTranslator.toRGB(hslBigHue, legacyOptions)).toBe(item.rgbLegacy);
            expect(ColorTranslator.toRGBObject(hslBigHue, options)).toMatchObject(item.rgbObject);

            expect(ColorTranslator.toHSL(hslBigHue, options)).toBe(item.hsl);
            expect(ColorTranslator.toHSL(hslBigHue, legacyOptions)).toBe(item.hslLegacy);
            expect(ColorTranslator.toHSLObject(hslBigHue, options)).toMatchObject(item.hslObject);

        });

        it(`HSLA big hue ${JSON.stringify(hslaBigHue)}`, (): void => {

            expect(ColorTranslator.toRGBA(hslaBigHue, options)).toBe(item.rgba);
            expect(ColorTranslator.toRGBA(hslaBigHue, legacyOptions)).toBe(item.rgbaLegacy);
            expect(ColorTranslator.toRGBAObject(hslaBigHue, options)).toMatchObject(item.rgbaObject);

            expect(ColorTranslator.toHSLA(hslaBigHue, options)).toBe(item.hsla);
            expect(ColorTranslator.toHSLA(hslaBigHue, legacyOptions)).toBe(item.hslaLegacy);
            expect(ColorTranslator.toHSLAObject(hslaBigHue, options)).toMatchObject(item.hslaObject);

        });

        it(`HSL low hue ${JSON.stringify(hslLowHue)}`, (): void => {

            expect(ColorTranslator.toRGB(hslLowHue, options)).toBe(item.rgb);
            expect(ColorTranslator.toRGB(hslLowHue, legacyOptions)).toBe(item.rgbLegacy);
            expect(ColorTranslator.toRGBObject(hslLowHue, options)).toMatchObject(item.rgbObject);

            expect(ColorTranslator.toHSL(hslLowHue, options)).toBe(item.hsl);
            expect(ColorTranslator.toHSL(hslLowHue, legacyOptions)).toBe(item.hslLegacy);
            expect(ColorTranslator.toHSLObject(hslLowHue, options)).toMatchObject(item.hslObject);

        });

        it(`HSLA low hue ${JSON.stringify(hslaLowHue)}`, (): void => {

            expect(ColorTranslator.toRGBA(hslaLowHue, options)).toBe(item.rgba);
            expect(ColorTranslator.toRGBA(hslaLowHue, legacyOptions)).toBe(item.rgbaLegacy);
            expect(ColorTranslator.toRGBAObject(hslaLowHue, options)).toMatchObject(item.rgbaObject);

            expect(ColorTranslator.toHSLA(hslaLowHue, options)).toBe(item.hsla);
            expect(ColorTranslator.toHSLA(hslaLowHue, legacyOptions)).toBe(item.hslaLegacy);
            expect(ColorTranslator.toHSLAObject(hslaLowHue, options)).toMatchObject(item.hslaObject);

        });

        it(`Hue in deg ${hslDeg}`, () => {

            expect(ColorTranslator.toRGB(hslDeg, options)).toBe(item.rgb);
            expect(ColorTranslator.toRGB(hslDeg, legacyOptions)).toBe(item.rgbLegacy);
            expect(ColorTranslator.toRGBObject(hslDeg, options)).toMatchObject(item.rgbObject);

            expect(ColorTranslator.toHSL(hslDeg, options)).toBe(item.hsl);
            expect(ColorTranslator.toHSL(hslDeg, legacyOptions)).toBe(item.hslLegacy);
            expect(ColorTranslator.toHSLObject(hslDeg, options)).toMatchObject(item.hslObject);

        });

        it(`Hue in grad ${hslGrad}`, () => {

            expect(ColorTranslator.toRGB(hslGrad, options)).toBe(item.rgb);
            expect(ColorTranslator.toRGB(hslGrad, legacyOptions)).toBe(item.rgbLegacy);
            expect(ColorTranslator.toRGBObject(hslGrad, options)).toMatchObject(item.rgbObject);

            expect(ColorTranslator.toHSL(hslGrad, options)).toBe(item.hsl);
            expect(ColorTranslator.toHSL(hslGrad, legacyOptions)).toBe(item.hslLegacy);
            expect(ColorTranslator.toHSLObject(hslGrad, options)).toMatchObject(item.hslObject);

        });

        it(`Hue in rad ${hslRadians}`, () => {

            expect(ColorTranslator.toRGB(hslRadians, options)).toBe(item.rgb);
            expect(ColorTranslator.toRGB(hslRadians, legacyOptions)).toBe(item.rgbLegacy);
            expect(ColorTranslator.toRGBObject(hslRadians, options)).toMatchObject(item.rgbObject);

            expect(ColorTranslator.toHSL(hslRadians, options)).toBe(item.hsl);
            expect(ColorTranslator.toHSL(hslRadians, legacyOptions)).toBe(item.hslLegacy);
            expect(ColorTranslator.toHSLObject(hslRadians, options)).toMatchObject(item.hslObject);

        });

        it(`Hue in turn ${hslTurns}`, () => {

            expect(ColorTranslator.toRGB(hslTurns, options)).toBe(item.rgb);
            expect(ColorTranslator.toRGB(hslTurns, legacyOptions)).toBe(item.rgbLegacy);
            expect(ColorTranslator.toRGBObject(hslTurns, options)).toMatchObject(item.rgbObject);

            expect(ColorTranslator.toHSL(hslTurns, options)).toBe(item.hsl);
            expect(ColorTranslator.toHSL(hslTurns, legacyOptions)).toBe(item.hslLegacy);
            expect(ColorTranslator.toHSLObject(hslTurns, options)).toMatchObject(item.hslObject);

        });

    });

});