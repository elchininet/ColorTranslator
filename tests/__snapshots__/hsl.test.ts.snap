// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HSL tests for hsl(0 0% 0%) HSL with percentages {"h":0,"s":"0%","l":"0%"} 1`] = `"rgb(0 0 0)"`;

exports[`HSL tests for hsl(0 0% 0%) HSL with percentages {"h":0,"s":"0%","l":"0%"} 2`] = `"hsl(0 0% 0%)"`;

exports[`HSL tests for hsl(0 0% 0%) HSLA with percentages {"h":0,"s":"0%","l":"0%","a":1} 1`] = `"rgb(0 0 0 / 1)"`;

exports[`HSL tests for hsl(0 0% 0%) HSLA with percentages {"h":0,"s":"0%","l":"0%","a":1} 2`] = `"hsl(0 0% 0% / 1)"`;

exports[`HSL tests for hsl(0 0% 50%) HSL with percentages {"h":0,"s":"0%","l":"50%"} 1`] = `"rgb(127.5 127.5 127.5)"`;

exports[`HSL tests for hsl(0 0% 50%) HSL with percentages {"h":0,"s":"0%","l":"50%"} 2`] = `"hsl(0 0% 50%)"`;

exports[`HSL tests for hsl(0 0% 50%) HSLA with percentages {"h":0,"s":"0%","l":"50%","a":1} 1`] = `"rgb(127.5 127.5 127.5 / 1)"`;

exports[`HSL tests for hsl(0 0% 50%) HSLA with percentages {"h":0,"s":"0%","l":"50%","a":1} 2`] = `"hsl(0 0% 50% / 1)"`;

exports[`HSL tests for hsl(0 0% 100%) HSL with percentages {"h":0,"s":"0%","l":"100%"} 1`] = `"rgb(255 255 255)"`;

exports[`HSL tests for hsl(0 0% 100%) HSL with percentages {"h":0,"s":"0%","l":"100%"} 2`] = `"hsl(0 0% 100%)"`;

exports[`HSL tests for hsl(0 0% 100%) HSLA with percentages {"h":0,"s":"0%","l":"100%","a":1} 1`] = `"rgb(255 255 255 / 1)"`;

exports[`HSL tests for hsl(0 0% 100%) HSLA with percentages {"h":0,"s":"0%","l":"100%","a":1} 2`] = `"hsl(0 0% 100% / 1)"`;

exports[`HSL tests for hsl(0 100% 50%) HSL with percentages {"h":0,"s":"100%","l":"50%"} 1`] = `"rgb(255 0 0)"`;

exports[`HSL tests for hsl(0 100% 50%) HSL with percentages {"h":0,"s":"100%","l":"50%"} 2`] = `"hsl(0 100% 50%)"`;

exports[`HSL tests for hsl(0 100% 50%) HSLA with percentages {"h":0,"s":"100%","l":"50%","a":1} 1`] = `"rgb(255 0 0 / 1)"`;

exports[`HSL tests for hsl(0 100% 50%) HSLA with percentages {"h":0,"s":"100%","l":"50%","a":1} 2`] = `"hsl(0 100% 50% / 1)"`;

exports[`HSL tests for hsl(120 100% 50%) HSL with percentages {"h":120,"s":"100%","l":"50%"} 1`] = `"rgb(0 255 0)"`;

exports[`HSL tests for hsl(120 100% 50%) HSL with percentages {"h":120,"s":"100%","l":"50%"} 2`] = `"hsl(120 100% 50%)"`;

exports[`HSL tests for hsl(120 100% 50%) HSLA with percentages {"h":120,"s":"100%","l":"50%","a":1} 1`] = `"rgb(0 255 0 / 1)"`;

exports[`HSL tests for hsl(120 100% 50%) HSLA with percentages {"h":120,"s":"100%","l":"50%","a":1} 2`] = `"hsl(120 100% 50% / 1)"`;

exports[`HSL tests for hsl(240 100% 50%) HSL with percentages {"h":240,"s":"100%","l":"50%"} 1`] = `"rgb(0 0 255)"`;

exports[`HSL tests for hsl(240 100% 50%) HSL with percentages {"h":240,"s":"100%","l":"50%"} 2`] = `"hsl(240 100% 50%)"`;

exports[`HSL tests for hsl(240 100% 50%) HSLA with percentages {"h":240,"s":"100%","l":"50%","a":1} 1`] = `"rgb(0 0 255 / 1)"`;

exports[`HSL tests for hsl(240 100% 50%) HSLA with percentages {"h":240,"s":"100%","l":"50%","a":1} 2`] = `"hsl(240 100% 50% / 1)"`;

exports[`HSL tests for hsl(300 100% 50%) HSL with percentages {"h":300,"s":"100%","l":"50%"} 1`] = `"rgb(255 0 255)"`;

exports[`HSL tests for hsl(300 100% 50%) HSL with percentages {"h":300,"s":"100%","l":"50%"} 2`] = `"hsl(300 100% 50%)"`;

exports[`HSL tests for hsl(300 100% 50%) HSLA with percentages {"h":300,"s":"100%","l":"50%","a":1} 1`] = `"rgb(255 0 255 / 1)"`;

exports[`HSL tests for hsl(300 100% 50%) HSLA with percentages {"h":300,"s":"100%","l":"50%","a":1} 2`] = `"hsl(300 100% 50% / 1)"`;
